### ISSUE TITLE
Auto-reconnect doesn't work as expected

### ISSUE THREAD

**Issue by UtechtDustin** [php-mqtt/client#129]
I tried to use the autoreconnect option (Implemented in #106 by @Namoshek), but i get a few exceptions and found some issues with the code.

My test script:
```php
$connectionSettings = (new \PhpMqtt\Client\ConnectionSettings())
    ->setUsername('username')
    ->setPassword('password')
    ->setReconnectAutomatically(true);


$last = 0;
$mqtt = new \PhpMqtt\Client\MqttClient('server', 'port', 'clientId');
$mqtt->registerLoopEventHandler(function  ($mqttClient, $elapsedTime) use (&$last) {
    if ($elapsedTime % 5 === 0 && $last !== (int)$elapsedTime) {
        $mqttClient->publish("xyz", 'true', 0, true);
        $last = (int)$elapsedTime;
    }
});

$mqtt->connect($connectionSettings);
$mqtt->loop();
$mqtt->close();
```

If i run this script and stop my broker i get this stacktrace
```
PHP Fatal error:  Uncaught PhpMqtt\Client\Exceptions\DataTransferException: [65] Transferring data over socket failed: Sending data over the socket failed. Has it been closed? in /var/www/html/vendor/php-mqtt/client/src/MqttClient.php:1143
Stack trace:
#0 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(1105): PhpMqtt\Client\MqttClient->writeToSocket()
#1 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(1074): PhpMqtt\Client\MqttClient->writeToSocketWithAutoReconnect()
#2 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(689): PhpMqtt\Client\MqttClient->ping()
#3 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(645): PhpMqtt\Client\MqttClient->loopOnce()
#4 /var/www/html/scripts/MqttService.php(41): PhpMqtt\Client\MqttClient->loop()
#5 {main}
  thrown in /var/www/html/vendor/php-mqtt/client/src/MqttClient.php on line 1143
```
This happens because [the client tries to ping the broke](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L687), but there is no open socket.

After that i tried to ignore the ping and deleted the ping call.
But now i get this stacktrace:
```
PHP Fatal error:  Uncaught TypeError: fread(): Argument #1 ($stream) must be of type resource, null given in /var/www/html/vendor/php-mqtt/client/src/MqttClient.php:1205
Stack trace:
#0 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(1205): fread()
#1 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(1263): PhpMqtt\Client\MqttClient->readFromSocket()
#2 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(672): PhpMqtt\Client\MqttClient->readAllAvailableDataFromSocket()
#3 /var/www/html/vendor/php-mqtt/client/src/MqttClient.php(645): PhpMqtt\Client\MqttClient->loopOnce()
#4 /var/www/html/scripts/MqttService.php(41): PhpMqtt\Client\MqttClient->loop()
#5 {main}
  thrown in /var/www/html/vendor/php-mqtt/client/src/MqttClient.php on line 1205
```
This happens because we try to read from the socket ([readFromSocket method](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1201)) which still don't exist.
Instead of the [readFromSocket method](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1201) the [readFromSocketWithAutoReconnect method](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1168) should be called i guess.
But this don't happen because the [readAllAvailableDataFromSocket method](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1168) has parameter called `$withAutoReconnectIfConfigured` which will never be set/used.
[On this part](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1259-L1262) the client always will use the [readFromSocket method](https://github.com/php-mqtt/client/blob/master/src/MqttClient.php#L1201).

**Comment by Namoshek** [comment]
What does the scenario you are trying to test look like? Are you simply restarting a (docker) service? Are you testing a failover with a load balancer between brokers? Or something else?

The current auto-reconnect implementation tries to reconnect once and throws the exception if reconnecting fails. This is useful for server-side timeouts as well as very basic failovers. It is not suitable for more complex scenarios like a 30second downtime of the broker.

**Comment by UtechtDustin** [comment]
Thank you for the really quick answer!

In our scenario we have a system which always needs to be connected with the mqtt broker.
Every 3 months we have to restart the docker container (which contains mosquitto) to replace the ssl certificate.
Another scenario would be a downtime of the server itself (which should not happen, but could as we all know :D).

So its a server-side downtime between 10-60 seconds.
But i thought this is the autoreconnect feature is for exactly that, i also woundered why there is no sleep between the reconnect tries.

**Comment by Namoshek** [comment]
The auto-reconnect is a bit of a difficult topic as it can be seen different for procuders and consumers. A producer may be a simple web application which publishes a message as part of a request, while a consumer is typically a long-running process (i.e. a console command). While the latter can easily wait a few seconds until a reconnect is possible, the former cannot (or should not) because it will cause high latency on the request.

Currently, the [reconnect logic](https://github.com/php-mqtt/client/blob/10750d8e0e82e24151e6cc5e960e4a32aa4161fa/src/MqttClient.php#L403-L425) is rather simple and only uses the configured attempts in combination with the connect timeout:
```php
$connectionSettings = (new \PhpMqtt\Client\ConnectionSettings)
    ->setConnectTimeout(60)
    ->setReconnectAutomatically(true)
    ->setMaxReconnectAttempts(3);
```


That said, I see that this may not work in all scenarios equally well and an additional configuration option for the delay between reconnect attempts would be good. I'll try to throw something together...

**Comment by UtechtDustin** [comment]
@Namoshek Thanks for the explanation.
The best case for a long-running process case would be if the phpmqtt client always tries to reconnect (no max attempts) with a delay between the reconnect tries.

Just a background information (maybe that helps to understand).
We use php-mqtt to run a systemd daemon, which is installed on a nuc (there will be a lot of systems).
Every 5 seconds we send an 'online' payload to the broker, to show the online/offline info on a webbased management page.
We also have a few topics which triggers actions on the nuc e.g. reboot, shutdown and a few more actions.

**Comment by Namoshek** [comment]
I understand that this may be useful in your case, but I doubt it is the ideal solution for everyone. However, adding a connection setting for an additional delay is of course not a problem and I did so in #130 - please have a look and give me your approval, if you think this looks alright!

Off-topic regarding your use case: not sure if you just left out this part, but wouldn't it make sense to use the _last will_ topic for the online/offline state?

**Comment by UtechtDustin** [comment]
I can test #130 tomorrow, but it looks good i guess.
That would fix the issue with the delay between the reconnects, but i'm not sure if it also fix the other issues (socket which is null while ping and try read data from socket - both stacktraces in the first post.)

We already use last will to tell the broker that the nuc is offline, but we still need the online topics for a few other things :-)

**Comment by Namoshek** [comment]
I think you are right about the other issue as well, there seems to be missing two things:
- The parameter `$withAutoReconnectIfConfigured` is never set to `true`.
- The parameter `$withAutoReconnectIfConfigured` indicates that even if it is set to `true`, the connection settings should be checked as well whether auto-reconnect is configured.

And another off-topic question regarding your use case: if you are using systemd as supervisor, can't you simply use the `Restart=on-failure` option to restart the PHP application, if the connection gets lost (and cannot be recovered)?
I mean, obviously it is better if the app can handle transient errors like a temporary connection loss. But ultimately, in the IoT world, you would want to have some kind of watchdog which ensures the app is restarted in case of failure.

**Comment by UtechtDustin** [comment]
We will use `Restart=on-failure`, but as you already said the the app itself should be stable as possible.
At the moment the restart option would also not help with the other two issues :)

**Comment by Namoshek** [comment]
Just to make sure we talk about the same expected behavior (which should be given with #130 and #131):
1. The client should auto-reconnect according to the connection settings during writes and reads up to the configured number of attempts with the configure delay between attempts.

2. The client will throw an exception indicating that writing/reading was not possible. If auto-reconnect is disabled, this exception is thrown immediately. If auto-reconnect is enabled, the exception is only thrown after all the configured attempts have been used. (If a connection could be re-established, no exception is thrown of course.)

**Comment by UtechtDustin** [comment]
That's exactly what i would expect, sounds great!
Thanks for the quick answers and PR's (and the great project at all).
This is why i love open source!

**Comment by Namoshek** [comment]
Alright, I published both PRs as [`v1.6.0`](https://github.com/php-mqtt/client/releases/tag/v1.6.0). I'll close this issue, but please let me know whether the changes helped. :+1:

**Comment by UtechtDustin** [comment]
Hey @Namoshek, i tested your changes.
It seems this working perfectly, thanks!
